extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java

```diff
@@ -246,4 +267,40 @@ public void stop() {
             }
         }
     }
+
+    /**
+     * Wrapper to facilitate filtering & sorting of applicable CDI produced AgroalDataSourceListeners
+     */
+    private class PrioritizedDataSourceListenerWrapper implements Comparable<PrioritizedDataSourceListenerWrapper> {
+        final AgroalDataSourceListener listener;
+        final int priority;
+        final String dataSourceName;
+
+        PrioritizedDataSourceListenerWrapper(final AgroalDataSourceListener listener, final int priority,
+                final String dataSourceName) {
+            this.listener = listener;
+            this.priority = priority;
+            this.dataSourceName = dataSourceName;
+        }
+
+        PrioritizedDataSourceListenerWrapper(final Bean<AgroalDataSourceListener> listener) {
+            this.listener = (AgroalDataSourceListener) beanManager.getReference(listener, listener.getBeanClass(),
+                    beanManager.createCreationalContext(listener));
+
+            Priority p = listener.getBeanClass().getAnnotation(Priority.class);
+            this.priority = p != null ? p.value() : Interceptor.Priority.APPLICATION;
+
+            io.quarkus.agroal.DataSource ds = listener.getBeanClass().getAnnotation(io.quarkus.agroal.DataSource.class);
+            this.dataSourceName = ds != null ? ds.value() : null;
+        }
+
+        boolean appliesTo(final String dataSourceNamed) {
+            return dataSourceName == null || dataSourceName.equals(dataSourceNamed);
+        }
+
+        @Override
+        public int compareTo(final PrioritizedDataSourceListenerWrapper o) {
+            return Integer.compare(this.priority, o.priority);
```

Comment:
we should also add a secondary comparison to ensure consistent ordering for the cases where two listeners are applied with the same priority. The secondary comparison could be considered an implementation detail and therefore doesn't need to be documented. Perhaps something like comparing on `listener.getClass().getCanonicalName()`?